astra/src/main/java/com/slack/astra/graphApi/GraphBuilder.java

```diff
@@ -0,0 +1,150 @@
+package com.slack.astra.graphApi;
+
+import com.slack.astra.zipkinApi.ZipkinSpanResponse;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * GraphBuilder constructs service dependency graphs from Zipkin span data.
+ *
+ * <p>This class processes distributed tracing spans to build a graph representation showing
+ * relationships between services. It creates nodes representing services and edges representing
+ * parent-child relationships between spans.
+ *
+ * <p>The builder supports configurable node metadata extraction through GraphConfig, allowing
+ * customization of which span tags are used to populate node metadata.
+ */
+public class GraphBuilder {
+  private static final Logger LOG = LoggerFactory.getLogger(GraphBuilder.class);
+  private final GraphConfig config;
+  private final Boolean defaultBehaviour;
+
+  /**
+   * Constructs a GraphBuilder with the specified configuration.
+   *
+   * @param config The GraphConfig to use for node metadata extraction. If GraphConfig.DEFAULT, uses
+   *     service name from a span's remote endpoint.
+   */
+  GraphBuilder(GraphConfig config) {
+    this.config = config;
+    this.defaultBehaviour = this.config == GraphConfig.DEFAULT;
+  }
+
+  /**
+   * Builds a dependency graph from a list of Zipkin spans.
+   *
+   * <p>This method processes spans to create nodes (services) and edges (dependencies) representing
+   * the service communication graph. Each span becomes a node, and parent-child relationships
+   * between spans become edges in the graph.
+   *
+   * @param spans List of Zipkin spans to process
+   * @return Graph containing nodes and edges representing service dependencies
+   */
+  public Graph buildFromSpans(List<ZipkinSpanResponse> spans) {
+    // Storage for unique nodes keyed by node ID
+    HashMap<String, Node> nodes = new HashMap<>();
+    // Mapping from span ID to the corresponding node ID
+    HashMap<String, String> spanIdToNodeId = new HashMap<>();
+    // Track parent-child relationships: parent span ID -> list of child node IDs
+    HashMap<String, ArrayList<String>> childrenByParentSpan = new HashMap<>();
+
+    // First pass: create nodes and build parent-child relationships
+    for (ZipkinSpanResponse span : spans) {
+      if (span.getId() == null) {
+        LOG.warn("Document={} cannot have missing id", span);
```

Comment:
maybe worded like this would be better, though I see you pulled this from the zipkin service's conversion, so maybe this would happen earlier in the stack anyway
```suggestion
        LOG.warn("Ignoring document with missing id, {}", span);
```